hail/python/hailtop/aiocloud/aioaws/fs.py

```diff
@@ -490,6 +499,8 @@ async def create(self, url: str, *, retry_writes: bool = True) -> S3CreateManage
         # complete before the write can begin (unlike the current
         # code, that copies 128MB parts in 256KB chunks).
         bucket, name = self.get_bucket_and_name(url)
+        if name == '':
+            raise IsABucketError(url)
```

Comment:
nit: I think I'd rather have this handled with a decorator (or at least a separate method). Is it bad DX to have these methods accept a `S3AsyncFSURL` and have a decorator on top tha converts `str` to the the URL object, optionally erroring if there's no path?

Reply:
So, my main gripe with decorators is that they tend to mangle the types. I tried writing a decorator with precise types [in this commit](https://github.com/hail-is/hail/commit/d3663075a67c2a4f44df600935a19d140c998299). I'm not sure why this fails:
```
pyright fs.py
/Users/dking/projects/hail/hail/python/hailtop/aiocloud/aioaws/fs.py
  /Users/dking/projects/hail/hail/python/hailtop/aiocloud/aioaws/fs.py:460:15 - error: "open" overrides method of same name in class "AsyncFS" with incompatible type "StrOrURLMethod[(), Coroutine[Any, Any, ReadableStream]]" (reportIncompatibleMethodOverride)
1 error, 0 warnings, 0 informations 
```

It works fine though. I can read stuff from S3.
```
In [1]: from hailtop.aiocloud.aioaws import S3AsyncFS
   ...: fs = S3AsyncFS()
   ...: (await (await fs.open('s3://gnomad-public-us-east-1/release/4.0/constraint/README.txt')).read())[:100]
Out[1]: b'\xef\xbb\xbfConstraint field descriptions\r\n\r\nDescriptions of columns in the gnomAD v4 constraint metrics tsv.'
```

Maybe I'm overcomplicating what you're asking, but there seems to be two issues:
1. If the URL is in the open method's signature either we make FS generic (ugh) or we assert that the URL is the right URL everywhere.
2. Preserving precise type information requires a fair bit of code (even if we can resolve (1)).

---

I was able to write a decorator that had no types and pyright didn't complain, but I also lost all type hinting on `fs.open`. IMO, between types and not repeating myself, I'd go with types.

Reply:
Let me try something using inheritance instead.

Reply:
Something along these lines may work and won't be too bad, but we'll need to play the same trick where the RouterAsyncFSURL has a reference to its file system. It will be a rather large change and I (1) would rather not tie that to this PR and (2) have committed to limiting my programming substantially in this last month. What do you think of leaving this as a follow up cleanup?

https://github.com/hail-is/hail/commit/d39557bf2bd099a6536bdc01640e5bad95c7810e

Reply:
hrmph, ya while I don't hate the last option in principal I'd rather not get that involved here. Can you just wrap the error throwing in a method so that most of these call-sites become one-liners and we can punt on the refactor?

Reply:
@daniel-goldstein OK, this was quite a few lines of code change, so worth a careful re-review.